Give user a chance to stop cancel event when there are changes

rlskoeser opened the pull request:
adds new logic to cancel event handler — if the currently active tiny editor has changes, open a confirmation prompt to give them a chance to opt out.

questions:
- should we have similar logic for changes to the annotation zone, in case of accidental cancellation?  (may matter more later when we have more complicated shapes)
- I would like to write a unit test, it doesn't look like we have one for the cancel handler right now — is that correct?

blms commented on src/index.ts:
```diff
@@ -100,6 +100,20 @@ class TranscriptionEditor {
+                // in the editor, give the user a chance to keep editing
+
+                // TODO: is there an equivalent check for annotation zone modified?
+                if (window.tinymce.activeEditor.isDirty()) {
```
Oh wow, great feature of tinymce!

blms reviewed (approved):
Looks good!

> should we have similar logic for changes to the annotation zone, in case of accidental cancellation? (may matter more later when we have more complicated shapes)

That might be a good idea, but you're right that it probably doesn't matter much now since it's so simple to redo with rectangles.

> I would like to write a unit test, it doesn't look like we have one for the cancel handler right now — is that correct?

Right, not for that `cancelSelected` event. (We do have some for the cancel button you might want to check to make sure nothing's being duplicated, though I doubt anything would be.)

blms reviewed (commented):
Code is looking good. But I just tested locally (Chrome Mac, with current `develop` branch of geniza active), and when I click "OK" to confirm that I want to continue editing, the selection still disappears in Annotorious. I can keep editing the text, but it doesn't persist when save is clicked. Let me know if you can't replicate and we can troubleshoot!

I'm also wondering if users might be tempted to click "cancel" without much thought if they didn't intend to trigger this confirmation dialog—but that would cancel their annotation! Kind of a tricky problem from a usability perspective, not sure whether "OK" to confirm cancellation or "Cancel" to cancel is better. But worth consideration I think.

rlskoeser commented:
@blms this is more complicated than I hoped. I added logic to re-select the annotation when we abort the cancel, but for some reason that fires the cancel event — so I added logic to skip cancellation when it's the annotation we're reselecting. 

Handling the cancel when selecting a different annotation block was even more complicated; I couldn't figure out a way to abort the cancel and get the right block selected, so I instead I made annotation blocks not clickable when one is being actively edited.  I am now able to to stop the cancellation and save my changes.

Ksenia confirmed that ideally we should prompt for confirmation when the cancel button is fired — but I think the confirm would need to be done in the annotation block cancel handler, not in the index code — the `handleCancel` method there seems too late. I wasn't sure of a good way to do that, so inclined to leave it out for now. (But please comment if you have suggestions.)

blms commented:
Thanks @rlskoeser, the code changes look great. Good call on the renaming, big improvements on disabling interactivity when editing, and that's wild about how reselection fires the cancel event again.

However, I'm still having trouble getting this to work (both in FF and Chrome Mac).

- Creating a new annotation: clicking outside of the selection removes the selection, but does not prompt for cancellation and does not do anything to the transcription block.
  - Clicking "save" then results in the "Saving..." alert to appear, but the network request is never made according to my network tab.
  - I am still allowed to create more annotations, so I can make a bunch of text blocks appear on the right by creating more and more selections on the canvas.
- Editing an existing annotation with modified text: clicking outside of the selection causes the cancellation prompt to show up
  - Clicking "OK" slightly changes the canvas zoom/pan (?) and reselects the selection on the canvas. When I then try to click "Save" on my annotation, I get the same behavior as above: "Saving..." alert, no network request.
  - Clicking "cancel" does indeed successfully cancel, as expected!
- Editing an existing annotation, without modifying text: successfully cancels both without prompt, as expected!

blms commented:
Here are some screen recordings in case the descriptions were unclear.

For what it's worth, there aren't any console errors or anything during either of these.

https://user-images.githubusercontent.com/7234006/197003789-46b98961-f0b4-445a-87c4-e91560c2c2e0.mov

https://user-images.githubusercontent.com/7234006/197004294-7f8224ec-f68c-48f9-8c81-c38283b1399b.mov

blms commented:
@rlskoeser I was able to get the `pointer-events` trick working like this. Pretty ugly, but I think it works from some brief testing. Want to give it a try along with any other changes you've made locally?

```js
    setAllInteractive(interactive: boolean) {
        if (interactive) {
            this.anno._element.firstElementChild.firstElementChild.style.pointerEvents = "auto";
            this.anno._element.querySelector("svg").style.pointerEvents = "auto";
        } else {
            this.anno._element.firstElementChild.firstElementChild.style.pointerEvents = "none";
            this.anno._element.querySelector("svg").style.pointerEvents = "none";
        }
    [...]
    }
```

blms commented:
@rlskoeser this one is ready for your review now!

rlskoeser commented on src/elements/CancelButton.ts:
```diff
@@ -39,6 +39,22 @@ class CancelButton extends HTMLButtonElement {
+                    "You have unsaved changes. Do you want to keep editing?",
+                ) == true
+            ) {
+                // if they click ok, return and don't process the cancelation
```
minor typo 
```suggestion
                // if they click ok, return and don't process the cancellation
```

rlskoeser reviewed (commented):
I tried to run this locally and I'm not seeing the new behavior. I pulled your changes, built tahqiq, installed local copy of tahqiq in geniza and built and ran with npm in the geniza project, but I'm not seeing the expected behavior. Did I miss a step? I'm running with the latest version of `develop` code.

I do see this error in the console: `t.parentNode is null`

rlskoeser commented:
Documenting my feedback from the slack convo: I tried again and was able to get the new code to run and see it in action. I think the inability to resize annotation zones is a deal-breaker, and I think we should get testing feedback on the two implementations of this code before we spend more time developing it. Maybe we could think about specific questions to ask that would help decide how to proceed.

@blms if you can set a version of the geniza codebase with a particular version of this branch, I can deploy. And then once we get a first round of testing feedback on one version, if you update I can deploy again for feedback on the other version.  (thinking the two versions are your latest changes and my last set of changes before that, which I think worked ok for all but new transcriptions, but didn't have the confirm logic on the cancel button).

rlskoeser commented:
Latest changes look reasonable, thanks for implementing.
